File: Python/025__Reverse_Nodes_in_K-Group.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        count = 0
        node = head

        while node and count < k:
            node = node.next
            count += 1
        if count < k:
            return head
        new_head, prev = self.reverse(head, count)
        head.next = self.reverseKGroup(new_head, k)

        return prev

    def reverse(self, head, count):

        prev = None
        cur = head
        next = head
        while count > 0:
            next = cur.next
            cur.next = prev
            prev = cur
            cur = next
            count -= 1
        return (cur, prev)

File: Java/025__Reverse_Nodes_in_k-Group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {

        ListNode rest = head;
        ListNode result = new ListNode(0);
        ListNode temp = null;
        ListNode prev_end = result;
        while (rest != null) {
            ListNode temp_result = null;
            ListNode cur_end = rest;
            int index = 0;
            temp = rest;
            while (temp != null) {
                temp = temp.next;
                index++;
            }
            if (index < k) {
                prev_end.next = rest;
                break;
            }
            index = 0;
            while (rest != null && index < k) {
                temp = temp_result;
                temp_result = rest;
                rest = rest.next;
                temp_result.next = temp;
                index++;
            }
            prev_end.next = temp_result;
            prev_end = cur_end;
        }

        return result.next;

    }
}

// A more structure Solution
class Solution {
  public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null) {
            return head;
        }
        ListNode tempHead = head, trailer = null;
        int count = 0;
        while (head != null && count < k) {
            trailer = head;
            head = head.next;
            count++;
        }
        if (head == null && count < k) {
            return tempHead;
        }
        trailer.next = null;
        ListNode result = reverseList(tempHead);
        tempHead.next = reverseKGroup(head, k);

        return result;
    }


    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode second = head.next;
        head.next = null;
        ListNode rest = reverseList(second);
        second.next = head;
        return rest;
    }
}
